### ISSUE TITLE
Performance improvements for workspace resolution

### ISSUE THREAD

**Issue by eseliger** [sourcegraph/sourcegraph#36260]
Workspace resolution is currently resource-hungry and can be slow for larger instances.

There are a few things I've identified that could be good improvements here:

- Can we change the `.batchignore` algorithm to always look at the default branch instead of the target branch?
  - If so, we can add a background worker that runs over all repos changed since the last run to persit the ignored status to a database table for fast lookups
- We can run the search from the worker itself. The search API has been refactored to run on any service now. This will help with resource contention on the frontend while the workspace resolver is running. 
  - There are currently two things that need to be solved before we can make it happen:
    - The logic to get user settings is still in the GQL layer, it would need to be factored out (good refactor in general)
    - The worker pod doesn't have all the env vars that are required to run all kinds of searches. Ideally it could get these config values from the frontend instead of us duplicating all the env vars across services. 
- Parallelize resolving `on` sections
- Use a batch method like [`BatchLog`](https://sourcegraph.com/github.com/sourcegraph/sourcegraph@es/race-condition-search-result-paths/-/blob/internal/gitserver/client.go?L667%3A13=) for `GetDefaultBranch`
- Investigate if we can use gitserver directly instead of search for `FindDirectoriesInRepos`, this runs for each repository. Ideally this also uses a batch method.

**Comment by courier-new** [comment]
Related to https://github.com/sourcegraph/sourcegraph/issues/43420.

**Comment by eseliger** [comment]
@courier-new did you want to split this into separate tickets or do I misremember? This was in sustaining before, so maybe 1 or 2 of the resulting tickets we might want to consider for the sustaining sprint?

**Comment by courier-new** [comment]
I saw it was in sustaining, so I wanted to leave it for us to talk about at sprint planning, to figure out which parts of it (if any) we want to tackle before splitting the issue up. 🙂 I figure we could split out the parts we want to work on into separate tickets and leave the rest here as a catch-all.

**Comment by courier-new** [comment]
Do you think you could help me with estimates for the items individually here, though, to prepare for tomorrow? 🙏

**Comment by courier-new** [comment]
Split out two separate issues for the two things we thing would have the biggest impact:
- https://github.com/sourcegraph/sourcegraph/issues/44569
- https://github.com/sourcegraph/sourcegraph/issues/44570

I'll label these remaining items as `strategic-ready/low`, and they will go to the backlog.
